### PR TITLE
run0: Implementation OCI process arguments (args).

### PR DESCRIPTION
**Pull request by kalil-pelissier**
Add the the functionality to run0 to handle the OCI process arguments for launching our container

### CODE REVIEW

**Comment by kalil-pelissier** [comment]
For the first implementation, run0 will handle only absolute path processes to avoid launching nested `sh` processes.
Observation: when launching a process like `ps` the RootfsCleanup fails with a "Broken pipe" error.

**Comment by kalil-pelissier** [comment]
I'm searching for how to handle command names only in the configuration file (*without an absolute path*) and find only one answer.  
All common binary files are located in the `/bin/` directory so I suggest prefixing all commands passed by the user by this path.    
> :-1: Probleme: If the container image is personalized with a binary file and added to a custom path this implementation will fail.

`config.json` file example:
```json
"process": {
                ...
                "args": [
                        "sh" 
                ]
                ...
}
```
@sameo what is your opinion on this implementation ?  
Will you suggest me another implementation ?

**Comment by sameo** [comment]
> @sameo what is your opinion on this implementation ? Will you suggest me another implementation ?

The `args` field is either absolute, or will be resolved by `execve` (see [here](https://pubs.opengroup.org/onlinepubs/9699919799/functions/exec.html) into an absolute path through the `PATH` environment variable.

So now that `run0` is applying the passed environment variable, you must not resolve `args` by yourself. Instead, `execve`/`clone` should do it for you, unless `PATH` is not passed through the config.json file.

**Comment by kalil-pelissier** [comment]
Thanks for the hints I finally understand how it should work :smile:.

I have updated my code for taking into consideration OCI arguments without an absolute path.
I also implement my code for the new Container Interface.

Ready to be reviewed @sameo.

File: run0/src/container/command.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+use oci_spec::runtime::Process;
+
+const UNSHARE_PROGRAM_ENTRYPOINT: &str = "/bin/sh";
```
Comment (sameo) [inline run0/src/container/command.rs]: We should not need that one.

File: run0/src/container/command.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+const DEFAULT_ARG0: &str = "sh";
```
Comment (sameo) [inline run0/src/container/command.rs]: Let's call it `DEFAULT_OCI_ARG` as it's an OCI specification.

File: run0/src/container/command.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+/// Implementation of the container arguments.
+#[derive(Default)]
+pub struct Command {
+    arg0: String,
+    args: Vec<String>,
+}
+
```
Comment (sameo) [inline run0/src/container/command.rs:10]: You want a `Default` implementation for `Command`, something that would set `arg0` to `sh` and `args` to an empty vector:

```rust
impl Default for Command {
    fn default() -> Self {
        Command {
            arg0: DEFAULT_OCI_ARG.to_string(),
            args: vec![],
        }
    }
}
```

File: run0/src/container/command.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+impl Command {
+    /// Get arguments.
+    /// It converts the `Self` into a `Command` for `unshare` API compatibility.
+    pub fn get(&self) -> unshare::Command {
```
Comment (sameo) [inline run0/src/container/command.rs]: What you're trying to do here is to convert a `Command` into an `unshare::Command`, so you should implement the `From` trait instead:

```rust
impl From<&Command> for unshare::Command {
}
```

File: run0/src/container/command.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        let mut command = unshare::Command::new(UNSHARE_PROGRAM_ENTRYPOINT);
```
Comment (sameo) [inline run0/src/container/command.rs]: And that one should just use whatever is set in `Command`:

```rust
let mut command: Command = unshare::Command::new(&self.arg0);
```

File: run0/src/container/command.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        command.arg0(&self.arg0).args(&self.args);
+
+        command
+    }
+}
+
+impl From<&Option<Process>> for Command {
+    fn from(process: &Option<Process>) -> Self {
```
Comment (sameo) [inline run0/src/container/command.rs:30]: Here  you would create a default `Command`, now that you have a proper `Default`  implementation for that structure.

```rust
let mut command: Command = Default::default();
```

And then modify the default values if there are some args from the spec.

File: run0/src/container/command.rs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        let mut args = Vec::<String>::new();
+        if let Some(process) = process {
+            if let Some(arguments) = process.args() {
+                for arg in arguments {
+                    args.push(arg.to_string());
+                }
```
Comment (sameo) [inline run0/src/container/command.rs]: This `for` loop is equivalent to `args = arguments.to_vec()`.

File: run0/src/container/command.rs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            }
+        }
+
+        let arg0 = match args.len() {
+            0 => String::from(DEFAULT_ARG0),
+            _ => args.remove(0),
+        };
+
```
Comment (sameo) [inline run0/src/container/command.rs:41]: Now that you start from a default Command, you would just need to check if `arguments.is_empty()` from the previous loop and do a `command.arg0 = command.args.remove(0);` if it is not.

File: run0/src/container/mod.rs [hunk 9]
```diff
@@ -79,7 +84,8 @@ impl Container {
     pub fn run(&self) -> Result<()> {
         let mounts = self.mounts.clone();
         let code = unsafe {
-            unshare::Command::new("/bin/sh")
+            self.command
+                .get()
```
Comment (sameo) [inline run0/src/container/mod.rs]: With a `From` implementation, this would rather be:

```rust
unshare::Command::from(&self.command)
```

**Review by sameo — COMMENTED** [review]
Thanks for the PR @kalil-pelissier. I have a few comments to be adressed.

Also, please rebase your PR on top of the latest main, we are now using a fork of the `unshare` crate that carries a fix for having libc resolving relative binary filenames with the `PATH` environment variables. That will make your PR work without the need to use `UNSHARE_PROGRAM_ENTRYPOINT` which is a absolute path.
See https://github.com/virt-do/unshare/commit/7b0a2e8d906bcdb2efa62d3ce9580dc359b731fd

**Comment by kalil-pelissier** [comment]
Thanks for the feedback @sameo.

I have rebased my PR and made all changes according to your comments.
Things are prettier when we are using rust built-in trait!

File: run0/src/container/command.rs [hunk 10]
```diff
@@ -0,0 +1,45 @@
+use oci_spec::runtime::Process;
+
+// const UNSHARE_PROGRAM_ENTRYPOINT: &str = "/bin/sh";
```
Comment (sameo) [inline run0/src/container/command.rs]: Leftover :)

**Review by sameo — COMMENTED** [review]
Thanks for the fixes @kalil-pelissier 
The code definitely looks simpler now and is ready to go in.

3 nits to fix before we can merge it:

1. You have one comment leftover (See my comment)
2. You have one linter to fix
3. Please squash your commits into a single one.

**Review by sameo — APPROVED** [review]

**Pull request merged**